text/0000-sandbox-environment.md

```diff
@@ -0,0 +1,276 @@
+what environment variables a piece of Rust code will access, and currently only
+has limited tools for tracking this.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+There are two unresolved questions in my mind:
+
+One is how to handle non-UTF-8 environment variable names and values? The
+standard library has `std::env::var_os` to fetch the environment in OS-encoded
+form, but there are no corresponing macros for compile-time. So I think just
+restricting the logical to pure UTF-8 for names and values is fine.
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+## Path handling
+[path-handling]: #path-handling
+
+Environment variables are frequently used for paths - a common pattern is:
+```
+include!(env!("SOME_PATH"));
+```
+
+If the variable `SOME_PATH` expands to a relative path, it is interpreted
+relative to the source file which contains the `include!()`. This requires the
+variable to be set with an explicit knowledge of which source files will
+reference the environment variable, and how they're layed out within the
+directory structure. If it is being using from multiple files in different
+direcctories (or specifically, at different directory depths), then no one
+relative path can be made to work. In practice the practical alternative is to
+always use absolute paths.
+
+To address this, an extension would be `--env-set-path VAR=PATH` (and
+`--env-pass-path VAR`) where the value is interpreted as a path relative to the
+`rustc` current working directory - in other words, it could be set as a
+relative path, but it would be interpreted as if it were an absolute path.
+Absolute paths would always be treated as absolute.
+
+## Tracked Environment API
+[tracked-env]: #tracked-env
+
+The [tracked environment API](https://github.com/rust-lang/rust/pull/74653)
+proposal adds an API to allow proc-macro accesses to the environment to be
+tracked appropriately. In the context of this RFC, these would be *directive*
+accesses, where the environment has a direct effect on the generate crate. As
+such it would access the logical environment proposed in this RFC.
+
+By the same token, if the proc-macro performs *operational* environment fetches,
+then it could still directly use the `std::env::var` API to fetch directly from
+the process environment. It is, of course, the proc-macro author's
+responsibility to understand the intent of their environment use, and correctly
+choose and use the appropriate API.
+
+## Regex Matching
+This proposal is intended to be a minimum set of functionality to control the
+environment. It requires very explicit control - aside from `--env-clear`, all
+variables must be explicitly listed to remove or set their values.
+
+A possible extension would be to allow regular expressions to select which
+variable names should be removed from the logical environment or passed through
+from the process environment. For example, `--env-remove-re` or `--env-pass-re`.
```

Comment:
I'd still prefer to make it explicit just to avoid any edge cases.

Also if these parameters can match multiple env vars, then you need to worry about how the options interact (eg, if you want to remove all the env vars matching a pattern except one, or conversely pass all except one). That will require some more specific design and not something that just a simple or obvious extension of naming a specific variable.